pydantic_ai_slim/pydantic_ai/ui/vercel_ai/_adapter.py

```diff
@@ -197,3 +215,226 @@ def load_messages(cls, messages: Sequence[UIMessage]) -> list[ModelMessage]:  #
+                        local_builtin_returns[part.tool_call_id] = part
+
+                for part in msg.parts:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        # Skip builtin tool returns - they're handled by the tool call logic
+                        continue
+                    elif isinstance(part, TextPart):
+                        # If this is the first text after an interruption, prepend separator
+                        if had_interruption:
+                            text_parts.append('\n\n' + part.content)
+                        else:
+                            text_parts.append(part.content)
+                    elif isinstance(part, ThinkingPart):
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+                            had_interruption = False
+                        ui_parts.append(ReasoningUIPart(text=part.content, state='done'))
+                    elif isinstance(part, FilePart):
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+                            had_interruption = False
+                        ui_parts.append(
+                            FileUIPart(
+                                url=part.content.data_uri,
+                                media_type=part.content.media_type,
+                            )
+                        )
+                    elif isinstance(part, BaseToolCallPart):  # pragma: no branch - All assistant part types are covered
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+
+                        # Mark that we had an interruption for next text part
+                        had_interruption = True
+
+                        if isinstance(part, BuiltinToolCallPart):
+                            prefixed_id = _make_builtin_tool_call_id(part.provider_name, part.tool_call_id)
+                            # Check local returns first (same message), then global returns (from ModelRequest)
+                            builtin_return = local_builtin_returns.get(part.tool_call_id) or (
+                                tool_returns.get(part.tool_call_id)
+                                if isinstance(tool_returns.get(part.tool_call_id), BuiltinToolReturnPart)
+                                else None
+                            )
+
+                            if builtin_return:
+                                content = builtin_return.model_response_str()
+                                call_provider_metadata = (
+                                    {'pydantic_ai': {'provider_name': part.provider_name}}
```

Comment:
Right, we were already doing this for builtin tools in the event stream already :)

I suppose the stuff I said above also applies to the other events we output like for reasoning. But since we get the data chunk by chunk and have to stream it using particular Vercel AI events, we may not always have the information a Vercel AI event needs at the time we get the Pydantic AI event. But for best behavior we really should use `provider_metadata` to store this kind of info there more as well. Feel free to look into that here or in a new PR.